pinot-query-planner/src/test/java/org/apache/calcite/rel/rules/PinotSortExchangeCopyRuleTest.java

```diff
@@ -0,0 +1,222 @@
+    Assert.assertTrue(((LogicalSort) sortCopy).getInput().getInput(0) instanceof LogicalSort);
+
+    LogicalSort innerSort = (LogicalSort) ((LogicalSort) sortCopy).getInput().getInput(0);
+    Assert.assertEquals(innerSort.getCollation().getKeys().size(), 0);
+    Assert.assertNull((innerSort).offset);
+    Assert.assertEquals((innerSort).fetch, literal(3));
+  }
+
+  @Test
+  public void shouldMatchSortOnly() {
+    // Given:
+    RelCollation collation = RelCollations.of(1);
+    SortExchange exchange = LogicalSortExchange.create(_input, RelDistributions.SINGLETON, collation);
+    Sort sort = LogicalSort.create(exchange, collation, null, null);
+    Mockito.when(_call.rel(0)).thenReturn(sort);
+    Mockito.when(_call.rel(1)).thenReturn(exchange);
+
+    // When:
+    PinotSortExchangeCopyRule.SORT_EXCHANGE_COPY.onMatch(_call);
+
+    // Then:
+    ArgumentCaptor<RelNode> sortCopyCapture = ArgumentCaptor.forClass(LogicalSort.class);
+    Mockito.verify(_call, Mockito.times(1)).transformTo(sortCopyCapture.capture(), Mockito.anyMap());
+
+    RelNode sortCopy = sortCopyCapture.getValue();
+    Assert.assertTrue(sortCopy instanceof LogicalSort);
+    Assert.assertTrue(((LogicalSort) sortCopy).getInput() instanceof LogicalSortExchange);
+    Assert.assertTrue(((LogicalSort) sortCopy).getInput().getInput(0) instanceof LogicalSort);
+
+    LogicalSort innerSort = (LogicalSort) ((LogicalSort) sortCopy).getInput().getInput(0);
+    Assert.assertEquals(innerSort.getCollation(), collation);
+    Assert.assertNull((innerSort).offset);
+    Assert.assertNull((innerSort).fetch);
+  }
+
+  @Test
+  public void shouldMatchLimitOffsetAndSort() {
+    // Given:
+    RelCollation collation = RelCollations.of(1);
+    SortExchange exchange = LogicalSortExchange.create(_input, RelDistributions.SINGLETON, collation);
+    Sort sort = LogicalSort.create(exchange, collation, literal(1), literal(2));
+    Mockito.when(_call.rel(0)).thenReturn(sort);
+    Mockito.when(_call.rel(1)).thenReturn(exchange);
+
+    // When:
+    PinotSortExchangeCopyRule.SORT_EXCHANGE_COPY.onMatch(_call);
+
+    // Then:
+    ArgumentCaptor<RelNode> sortCopyCapture = ArgumentCaptor.forClass(LogicalSort.class);
+    Mockito.verify(_call, Mockito.times(1)).transformTo(sortCopyCapture.capture(), Mockito.anyMap());
+
+    RelNode sortCopy = sortCopyCapture.getValue();
+    Assert.assertTrue(sortCopy instanceof LogicalSort);
+    Assert.assertTrue(((LogicalSort) sortCopy).getInput() instanceof LogicalSortExchange);
+    Assert.assertTrue(((LogicalSort) sortCopy).getInput().getInput(0) instanceof LogicalSort);
+
+    LogicalSort innerSort = (LogicalSort) ((LogicalSort) sortCopy).getInput().getInput(0);
+    Assert.assertEquals(innerSort.getCollation(), collation);
+    Assert.assertNull((innerSort).offset);
+    Assert.assertEquals((innerSort).fetch, literal(3));
+  }
+
+  @Test
+  public void shouldNotMatchOnlySortAlreadySorted() {
```

Comment:
seems like alreadySorted is tested but  alreadySmaller is not. from 
```
public static boolean checkInputForCollationAndLimit( 
 ... {
  return alreadySorted(mq, input, collation) && alreadySmaller(mq, input, offset, fetch);
}
```
but given we are dealing with exchange / sort, i assume already smaller is always satisfied. yes?

Reply:
this is much harder to test, essentially the "alreadySmaller" could see that there is already an inner sort with a limit that limits the rows to fewer than what we'd need to satisfy the sort in question - alternatively, we could have table stats to show that the input is already smaller than the limit in question.